### PR TITLE
Multiphop support for CurveV1Factory and CurveV1StableNg on v6 + buy support for CurveV1Factory factoryPlain2CrvEma pools

### PR DESCRIPTION
**Pull request by aburkut**

### CODE REVIEW

File: src/dex/curve-v1-factory/curve-v1-factory.ts [hunk 1]
```diff
@@ -661,6 +672,8 @@ export class CurveV1Factory
               basePoolStateFetcher,
               factoryImplementationFromConfig.customGasCost,
               factoryImplementationFromConfig.isStoreRateSupported,
+              undefined,
+              undefined,
```
Comment (KanievskyiDanylo) [inline src/dex/curve-v1-factory/curve-v1-factory.ts:676]: No need to pass `undefined`

File: src/dex/curve-v1-stable-ng/curve-v1-stable-ng.ts [hunk 2]
```diff
@@ -7,6 +7,7 @@ import { IDexHelper } from '../../dex-helper';
 import { Adapters, CurveV1StableNgConfig } from './config';
 import { AbiItem } from 'web3-utils';
 import { getDexKeysWithNetwork } from '../../utils';
+import { FactoryImplementationNames } from '../curve-v1-factory/types';
```
Comment (KanievskyiDanylo) [inline src/dex/curve-v1-stable-ng/curve-v1-stable-ng.ts:10]: redundant import

File: src/dex/curve-v1-factory/curve-v1-factory.ts [hunk 3]
```diff
@@ -1175,31 +1206,91 @@ export class CurveV1Factory
...
+      while (swapParams.length < swapParamsLength) {
+        swapParams.push([0, 0, 0, 0, 0]);
+      }
+
+      const pools = [];
+
+      while (pools.length < poolsLength) {
+        pools.push(NULL_ADDRESS);
+      }
+
```
Comment (KanievskyiDanylo) [inline src/dex/curve-v1-factory/curve-v1-factory.ts:1269]: @aburkut could you clarify this part? not clear why we need to apply such logic to build transaction

File: src/dex/curve-v1-factory/curve-v1-factory.ts [hunk 4]
```diff
@@ -1175,31 +1206,91 @@ export class CurveV1Factory
...
-      targetExchange: exchange,
-      returnAmountPos: undefined,
-    };
+      return {
+        exchangeData,
+        needWrapNative: this.needWrapNative,
+        sendEthButSupportsInsertFromAmount: true,
+        dexFuncHasRecipient: false,
+        targetExchange: exchange,
+        returnAmountPos: undefined,
```
Comment (0xNazarii) [inline src/dex/curve-v1-factory/curve-v1-factory.ts:1231]:
```suggestion
```

**Review by KanievskyiDanylo — APPROVED** [review]

**Review by 0xNazarii — APPROVED** [review]

File: src/dex/curve-v1-factory/curve-v1-factory.ts (hunk 1 above)
Comment (aburkut) [reply src/dex/curve-v1-factory/curve-v1-factory.ts:676]: @KanievskyiDanylo fixed

File: src/dex/curve-v1-stable-ng/curve-v1-stable-ng.ts (hunk 2 above)
Comment (aburkut) [reply src/dex/curve-v1-stable-ng/curve-v1-stable-ng.ts:10]: @KanievskyiDanylo fixed

File: src/dex/curve-v1-factory/curve-v1-factory.ts (hunk 3 above)
Comment (aburkut) [reply src/dex/curve-v1-factory/curve-v1-factory.ts:1269]: @KanievskyiDanylo Posted a comment with a link to Curve NG Router exchange function's params description

File: src/dex/curve-v1-factory/curve-v1-factory.ts [hunk 5]
```diff
@@ -1175,31 +1206,91 @@ export class CurveV1Factory
...
+      };
+    } else {
```
[the remark this answers is not in the document]
Comment (aburkut) [reply src/dex/curve-v1-factory/curve-v1-factory.ts:1233]: @0xNazarii refactored this part.

**Pull request merged**